app/src/main/kotlin/edu/stanford/spezi/app/onboarding/DefaultConsentRepository.kt

```diff
@@ -0,0 +1,35 @@
+package edu.stanford.spezi.app.onboarding
+
+import edu.stanford.spezi.core.coroutines.di.Dispatching
+import edu.stanford.spezi.core.navigation.DefaultNavigationEvent
+import edu.stanford.spezi.core.navigation.Navigator
+import edu.stanford.spezi.module.onboarding.consent.ConsentData
+import edu.stanford.spezi.module.onboarding.consent.ConsentRepository
+import edu.stanford.spezi.module.onboarding.consent.PdfService
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+class DefaultConsentRepository @Inject internal constructor(
+    private val pdfService: PdfService,
+    private val navigator: Navigator,
+    @Dispatching.IO private val ioScope: CoroutineScope,
+) : ConsentRepository {
+
+    override suspend fun getConsentData(): ConsentData {
```

Comment:
https://github.com/StanfordSpezi/SpeziKt/pull/12#discussion_r1615311593

i had originally split it because it was split in the iOS, but actually the difference is minimal. it is still split in the object itself and visually it is even more similar to the iOS version of it when the data is in an single object, because with us the object is then injected. i could also imagine that this makes it easier to react to errors if, for example, there are problems retrieving the data... which could theoretically also be loaded from a database or similar using the suspend function and are always required together. and as @eldcn says, it is probably rather unlikely that the data will come from different sources, but this is still technically possible in the current implementation @PSchmiedmayer 


as reference currently: 
iOS

`OnboardingConsentView(
            markdown: {
                Data("This is a *markdown* **example**".utf8)
            },
            action: {
                // Action to perform once the user has given their consent
            },
            exportConfiguration: .init(paperSize: .usLetter)    // Configure the properties of the exported consent form
        )`


vs Android

`ConsentData(
            markdownText = """
       This is a *markdown* **example**
    """,
            onAction = {
                 // Action to perform once the user has given their consent
            },
            }
        )`

Reply:
Good points; I would leave this up to you and @eldcn to decide. The main concern is mainly due to the nested nature itself. While on iOS you directly pass in the parameters in the view itself here you would (1) generate a repository and (2) within the repository create an object that handles that interaction instead of only (1) generate a repository that contains these values right out of the box. Both is fine with me and I would defer this to @eldcn and ideally some example from how Google defines some more complex views as part of their compose libraries and how they handle information injected into views.

I have so far checked out these guidelines on building components in Compose and they all don't seem to introduce a Repository or elements for this for lower-level component like individual views: https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md. Even more complicated views like these ones here in the docs have dedicated "onClick" functions passed in the constructors rather then offloading this to a dedicated type: https://developer.android.com/develop/ui/compose/mental-model. 

Looking at Google Frameworks like maps, they seem to follow a similar approach to SwiftUI where elements are all passed in through the initializer: https://googlemaps.github.io/android-maps-compose/maps-compose/com.google.maps.android.compose/-google-map.html. Arguably a lot of arguments but I think this makes sense for a good amount of basic building blocks. A simpler example is e.g., https://github.com/google/accompanist/blob/main/drawablepainter/src/main/java/com/google/accompanist/drawablepainter/DrawablePainter.kt.

More complicated elements might require elements where these things are extracted into dedicated types that can contain states but we reserve this for a few limited use cases where we consider things to be a complete "module" on the Swift side.
Would be great to identify where we establish a good middle ground and at what point we should move to repositories while some basic views could be configured similar to fundamental compose building blocks ...

CC @pauljohanneskraft, @eldcn, and @vishnuravi here as well.

Reply:
Many thanks for the feedback. I have looked at the examples and realised that they are mostly single composeables and not whole views. One of the reasons why we are currently proceeding in this way is the use of MVVM. We use the UiState from the ViewModel in the view. If we were to handle it like iOS, we would have to enter the data for the ViewModel in the view and then forward it to the ViewModel, even though the view does not need the data. The ViewModel would then return the state to the view. In my opinion, this does not quite correspond to the clean implementation of MVVM.

To get the data into the ViewModel, we have to inject it there in some form. Technically, we could also inject the ConsentData object directly, for example, and make it available in one place. However, this would limit us due to the static nature compared to the current dynamic provision.

Another disadvantage would be that we could not implement this in a standardised way. Especially when loading data from the UI, e.g. from an API, I think this would have to be done in a suitable service in the ViewModel anyway. I look forward to hearing the opinions of others.

For simpler `Composable` we currently do it similar to the linked sources above from you, @PSchmiedmayer 

`@Composable
fun ValidatedOutlinedTextField(
    modifier: Modifier = Modifier,
    value: String = "",
    onValueChange: (String) -> Unit,
    labelText: String = "",
    errorText: String? = null,
) {`

Reply:
Hey @PSchmiedmayer these are some great arguments that you are bringing here, let me share my perspective on these thoughts

> I am wondering what's the reasoning at to make this a single type that needs to be provided compared to having a requirement to have a property called markdown text and one that is a function that could be called on pressing the consent button?

I wont to refer to my comment [here](https://github.com/StanfordSpezi/SpeziKt/pull/12#discussion_r1615311593) that @Basler182 has also mentioned. In that example, we were solely providing static content (and no invokable actions from ui layer) via three methods of the repository, one method title, subtitle and list of areas. My proposal was encapsulate these three methods in a single data class object, due to the reasons that I have mentioned in the comment. Since here we also want have some actions to invoke, my suggestion (and also personal preference) is to define the actions as interface requirements instead of lamda (closure) properties of a single type. I already applied this in [EngageConsentManager](https://github.com/StanfordSpezi/SpeziKt/pull/20/files#diff-6c6f8439fa7f562c077fecd84cb5c046e61cab78bc2de3bc28e53778486ec971), and also added a todo for us to clarify for the remaining components in [EngageInvitationCodeRepository](https://github.com/StanfordSpezi/SpeziKt/pull/20/files#diff-07375133757e99947d960c2beef9bfa945b6e114e167d9b4cdc406e533c59888). The following open points for us to discuss:
- If we would stick with defining actions as interface requirements, it would make sense to revisit naming of these components, as they are not repositories anymore, but rather some managers, services, handlers of a certain module that should be provided via the consumer apps in order to use a spezi module.
- We are currently assuming that the data of these repositories is static. Do we need to support any reactive approach there? E.g. exposing StateFlow `fun getMarkdownText(): StateFlow<String>`?

> The main concern is mainly due to the nested nature itself. While on iOS you directly pass in the parameters in the view itself here you would (1) generate a repository and (2) within the repository create an object that handles that interaction instead of only (1) generate a repository that contains these values right out of the box.

I think in this case we should consider the nature of a spezi module itself:
1. Does the module provide only some Views (Composables) that the apps can just use in their screens? If yes, then there is for sure no need to have the requirement for the repositories, as the apps can invoke the module composable with the parameter values that they need.
2. Does the module perform additional functionalities besides of solely providing a Composable with static content, e.g. the `modules:account`? In this case I think the best approach is to have the repository or service requirement interface as the spezi module does much more under the hood. With this approach we leave less work for the apps as the whole setup is handle via the spezi module itself.

Considering the modules that we have introduced so far (account, onboarding and contact), I am more in favor of the repositories approach, as from my perspective those aren't that simple views .

> Both is fine with me and I would defer this to @eldcn and ideally some example from how Google defines some more complex views as part of their compose libraries and how they handle information injected into views.

All views provided via [compose.material3](https://developer.android.com/jetpack/androidx/releases/compose-material3) are meant for general purpose and fall into the first point above. Hence, I don't think some compose examples are the best reference when dealing with complex views. Compose views only serve as building blocks for entire screens.

To answer @PSchmiedmayer's question on how to deal with more complex views, I want to refer to [Circuit](https://slackhq.github.io/circuit/), which is an architecture from Slack to build reactive complex screens with Compose. I have tried to apply a way more simplified architecture in our modules so far which is also based on ui states and events